library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java

```
@@ -384,20 +394,19 @@ public void put(Slice key, long timeBucket, Slice value)
     {
       BucketedValue bucketedValue = flash.get(key);
       if (bucketedValue == null) {
-        bucketedValue = new BucketedValue();
+        bucketedValue = new BucketedValue(timeBucket, value);
         flash.put(key, bucketedValue);
-        sizeInBytes.getAndAdd(key.length);
-        sizeInBytes.getAndAdd(Long.SIZE);
-      }
-      if (timeBucket >= bucketedValue.getTimeBucket()) {
-
-        int inc = null == bucketedValue.getValue() ? value.length : value.length - bucketedValue.getValue().length;
-        sizeInBytes.getAndAdd(inc);
-        bucketedValue.setTimeBucket(timeBucket);
-        bucketedValue.setValue(value);
+        sizeInBytes.getAndAdd(key.length + value.length + Longs.BYTES);
       } else {
-        LOG.warn("ignoring {} {} {}; existing {} {}", key, value, timeBucket,
-            bucketedValue.getValue(), bucketedValue.getTimeBucket());
+        if (timeBucket >= bucketedValue.getTimeBucket()) {
+
+          int inc = null == bucketedValue.getValue() ? value.length : value.length - bucketedValue.getValue().length;
+          sizeInBytes.getAndAdd(inc);
+          bucketedValue.setTimeBucket(timeBucket);
+          bucketedValue.setValue(value);
+        } else {
+          throw new RuntimeException("newer entry exists for " + key);
```

Nit: use AssertionError instead?

---

I think an exception is better since the caller may want to catch it when that happens

---

I added an assert statement. 
@davidyan74 IMO handling these scenarios by catching an exception is costly. Instead, the caller should cache the value of key and time if they care about it.

---

The problem with the assert statement is that it is turned off by default. So if we use it here, then we are back to the silently ignored put. IMO it is better to explicitly raise an error in this situation.

@davidyan74 You can catch AssertionError just as you can catch RuntimeException. But it won't provide much value for handling the situation.